src/common/connection.ts

```diff
@@ -76,27 +152,147 @@ class Ledger {
+      this.promisesAwaitingConnection.push({resolve, reject})
+    })
+  }
+}
+
+/**
+ * Manage all the requests made to the websocket, and their async responses
+ * that come in from the WebSocket. Responses come in over the WS connection
+ * after-the-fact, so this manager will tie that response to resolve the
+ * original request.
+ */
+class RequestManager {
+  private nextId = 0
+  private promisesAwaitingResponse: {
+    resolve: Function
+    reject: Function
+    timer: NodeJS.Timeout
+  }[] = []
+
+  cancel(id: number) {
+    const {timer} = this.promisesAwaitingResponse[id]
+    clearTimeout(timer)
+  }
+
+  resolve(id: number, data: any) {
+    const {timer, resolve} = this.promisesAwaitingResponse[id]
+    clearTimeout(timer)
+    resolve(data)
+  }
+
+  reject(id: number, error: Error) {
+    const {timer, reject} = this.promisesAwaitingResponse[id]
+    clearTimeout(timer)
+    reject(error)
+  }
+
+  rejectAll(error: Error) {
+    this.promisesAwaitingResponse.forEach((_, id) => {
+      this.reject(id, error)
+    })
+  }
+
+  /**
+   * Creates a new WebSocket request. This sets up a timeout timer to catch
+   * hung responses, and a promise that will resolve with the response once
+   * the response is seen & handled.
+   */
+  createRequest(data: any, timeout: number): [number, string, Promise<any>] {
+    const newId = this.nextId++
+    const newData = JSON.stringify({...data, id: newId})
+    const timer = setTimeout(
+      () => this.reject(newId, new TimeoutError()),
+      timeout
+    )
+    // Node.js won't exit if a timer is still running, so we tell Node to ignore.
+    // (Node will still wait for the request to complete).
+    if (timer.unref) {
+      timer.unref()
+    }
+    const newPromise = new Promise((resolve, reject) => {
+      this.promisesAwaitingResponse[newId] = {resolve, reject, timer}
+    })
+    return [newId, newData, newPromise]
+  }
+
+  /**
+   * Handle a "response" (any message with `{type: "response"}`). Responses
+   * match to the earlier request handlers, and resolve/reject based on the
+   * data received.
+   */
+  handleResponse(data: any) {
+    if (!Number.isInteger(data.id) || data.id < 0) {
+      throw new ResponseFormatError('valid id not found in response', data)
+    }
+    if (!this.promisesAwaitingResponse[data.id]) {
+      throw new ResponseFormatError('response handler not found', data)
+    }
+    if (data.status === 'error') {
+      const error = new RippledError(data.error_message || data.error, data)
+      this.reject(data.id, error)
+      return
+    }
+    if (data.status !== 'success') {
+      const error = new ResponseFormatError(
+        `unrecognized status: ${data.status}`,
+        data
+      )
+      this.reject(data.id, error)
+      return
+    }
+    this.resolve(data.id, data.result)
+  }
+}
+
+/**
+ * The main Connection class. Responsible for connecting to & managing
+ * an active WebSocket connection to a Ripple node.
```

Comment:
Ripple node -> XRPL node

Reply:
another reason for doing that major release that we talked about: we could finally rename from ripple-lib to xrpl-lib :)